apps/aecore/src/pow_sha256.erl

```diff
@@ -0,0 +1,105 @@
+%%%=============================================================================
+%%% @copyright 2017, Aeternity Anstalt
+%%% @doc
+%%%    Module dealing with SHA256 Proof of Work calculation and hash generation
+%%% @end
+%%%=============================================================================
+-module(pow_sha256).
+
+-export([generate/3,
+         recalculate_difficulty/3]).
+
+-ifdef(TEST).
+-compile(export_all).
+-include_lib("eunit/include/eunit.hrl").
+-endif.
+
+-include("sha256.hrl").
+
+-define(NONCE_RANGE, 1000000000000000000000000).
+
+%% Integers encrypted in scientific notation:
+%%   256 * <base-2 exponent> + the most significant byte (i.e., the significand)
+-type sci_int() :: integer().
+
+-export_type([sci_int/0]).
+
+
+%%%=============================================================================
+%%% API
+%%%=============================================================================
+
+%%------------------------------------------------------------------------------
+%% Generate a nonce value that, when used in hash calculation of 'Data', results in
+%% a smaller value than the difficulty. Return 'not_found' if the condition is still
+%% not satisdfied after trying 'Count' times with incremented nonce values.
+%%------------------------------------------------------------------------------
+-spec generate(Data :: sha256:hashable(), Difficulty :: sci_int(), Count :: integer()) ->
+                      integer() | 'not_found'.
```

Comment:
Please prefer as return value:
* Either a tuple `{error, Reason::any()}`
* Or a tuple with at least two elements where the first element is `ok`.

Reply:
(of course you can put a more specific `Reason` in this case.)

Reply:
+1